api/interceptor/observer.go

```diff
@@ -0,0 +1,269 @@
+package interceptor
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"time"
+
+	"connectrpc.com/connect"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+)
+
+var _ connect.Interceptor = &Observer{}
+
+// Observer wraps all requests/responses and emits events that inform about
+// request duration, size and status code.
+type Observer struct {
+	onRequestEnded func(context.Context, *RequestMetadata)
+}
+
+// NewObserver creates a new Observer.
+func NewObserver(onRequestEnded func(context.Context, *RequestMetadata)) *Observer {
+	return &Observer{
+		onRequestEnded: onRequestEnded,
+	}
+}
+
+// WrapHandler mounts an HTTP middleware to gather metrics at the
+// HTTP level.
+func (in *Observer) WrapHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// 1. Inject request metadata into request context
+		var procedure string
+		switch {
+		case strings.HasPrefix(r.RequestURI, "/grpc.reflection.v1.ServerReflection"),
+			strings.HasPrefix(r.RequestURI, "/grpc.reflection.v1alpha.ServerReflection"):
```

Comment:
I am not sure I understand this check? Can you elaborate on it please.

Reply:
Tools like grpcurl or postman use grpc reflection to discover the gRPC endpoints. For each endpoint one request will be sent. If we don't have this case no procedure  (defaults to `none`) will be set.